main/src/main/java/org/mobilitydata/gtfsvalidator/validator/StopTooFarFromTripShapeValidator.java

```diff
@@ -47,32 +58,100 @@
+        // No shape for this trip - skip to the next trip
+        continue;
+      }
+      // Check for possible errors for this combination of stop times and shape points for
+      // this trip_id
+      List<StopTooFarFromTripShapeNotice> noticesForTrip =
+          checkStopsWithinTripShape(
+              tripId,
+              stopTimesForTrip,
+              trip.shapeId(),
+              shapeTable.byShapeId(trip.shapeId()),
+              stopTable,
+              testedCache);
+      for (StopTooFarFromTripShapeNotice notice : noticesForTrip) {
+        noticeContainer.addValidationNotice(notice);
+      }
+    }
+  }
+
+  /**
+   * Returns a list of notices for the given input, one for each stop that is too far from the trip
+   * shapePoints
+   *
+   * @param tripId trip_id for this GTFS trip
+   * @param stopTimes a list of StopTimes for a trip, sorted by stop_sequence
+   * @param shapeId the shape_id for this GTFS trip
+   * @param shapePoints a list of ShapePoints for a trip, sorted by shape_pt_sequence
+   * @param stopTable a container for all stops (keyed on stop_id), needed to obtain the latitude
+   *     and longitude for each stop on the trip
+   * @param testedCache a cache for previously tested shape_id and stop_id pairs (keyed on
+   *     shape_id+stop_id). If the combination of shape_id and stop_id appears in this set, we
+   *     shouldn't test it again. Shapes and stops tested in this method execution will be added to
+   *     this testedCache.
+   * @return a list of notices, one for each stop that is too far from the trip shape
+   */
+  List<StopTooFarFromTripShapeNotice> checkStopsWithinTripShape(
+      final String tripId,
+      final List<GtfsStopTime> stopTimes,
+      final String shapeId,
+      final List<GtfsShape> shapePoints,
+      final GtfsStopTableContainer stopTable,
+      final Set<String> testedCache) {
+    List<StopTooFarFromTripShapeNotice> notices = new ArrayList<>();
+    if (stopTimes == null || stopTimes.isEmpty() || shapePoints == null || shapePoints.isEmpty()) {
+      // Nothing to do - return empty list
+      return notices;
+    }
+
+    // Create a polyline from the GTFS shapes data
+    ShapeFactory.LineStringBuilder lineBuilder = GeospatialUtil.getShapeFactory().lineString();
+    for (GtfsShape shapePoint : shapePoints) {
+      lineBuilder.pointXY(shapePoint.shapePtLon(), shapePoint.shapePtLat());
+    }
+    Shape shapeLine = lineBuilder.build();
+
+    // Create the buffered version of the trip as a polygon
+    Shape shapeBuffer = shapeLine.getBuffered(TRIP_BUFFER_DEGREES, shapeLine.getContext());
+
+    // Check if each stop is within the buffer polygon
+    for (GtfsStopTime stopTime : stopTimes) {
+      GtfsStop stop = stopTable.byStopId(stopTime.stopId());
+      if (stop == null || !stop.hasStopLat() || !stop.hasStopLon()) {
+        // Lat/lon are optional for location_type 4 - skip to the next stop if they aren't
```

Comment:
We should use the coordinates of the parent location (i.e., platform) if stop_lat or stop_lon are missing.

Reply:
Any suggestions on how to report that the parent stop has been checked in the notice? Should we have a separate `ParentStopTooFarFromTripShapeNotice`? Or a new field in the notice map?

Reply:
Don't we risk to generate twice the same notice? One once the parent is checked, and the second one when the children is checked?

Reply:
>Don't we risk to generate twice the same notice: one once the parent is checked, and the second one when the child is checked?

Good point - we would if we didn't track that we generate a notice for this parent and we encounter multiple children that don't have lat/longs for the same parent. But we could do something similar to what we do now, and track via the cache that we already generated a notice for this parent & shape combination.

Reply:
So it looks like the spec was clarified since I first wrote this rule (https://github.com/google/transit/pull/258). It now states:

>Referenced locations must be stops/platforms, i.e. their stops.location_type value must be 0 or empty.

And given that `stop_lat` and `stop_lon` are required for `location_type=0` stops, this is actually a non-issue - we should never encounter stops in stop_times.txt without locations, and if we do we can ignore them.